File: day-14.py
from collections import Counter
from operator import methodcaller as mc
from typing import Dict, Tuple

from packages.aoc_helper import Input


def readto_string_dict() -> Tuple[str, Dict[str, str]]:
    _input = Input.readto_list("./data/d14")
    _starter, _ = _input.pop(0), _input.pop(0)
    _insertion_rules = dict(map(mc("split", " -> "), _input))
    return _starter, _insertion_rules


def max_minus_min(mono_counter: Counter) -> int:
    _values = mono_counter.values()
    return max(_values) - min(_values)


def part1and2(poly_starter_rules: Tuple[str, Dict[str, str]], steps: int) -> int:
    monomer_counter = Counter(poly_starter_rules[0])
    polymer_counter = Counter(
        map(
            lambda monomer: monomer[0] + monomer[1],
            zip(poly_starter_rules[0], poly_starter_rules[0][1:]),
        )
    )
    for _ in range(steps):
        tmp_counter = Counter()
        for key, value in polymer_counter.items():
            tmp_counter[key[0] + poly_starter_rules[1][key]] += value
            tmp_counter[poly_starter_rules[1][key] + key[1]] += value
            monomer_counter[poly_starter_rules[1][key]] += value
        polymer_counter = tmp_counter
    return max_minus_min(monomer_counter)


if __name__ == "__main__":
    print(f"Part 1: {part1and2(readto_string_dict(), 10)}")
    print(f"Part 2: {part1and2(readto_string_dict(), 40)}")

File: day-14.rb
def readto_string_hash(path="./data/d14")
    data = File.open(path).read.split("\n")
    starter = data.shift
    data.shift

    insertion_rules = {}
    for idx in 0..(data.length() - 1)
        insertion_rules[data[idx][0..1]] = data[idx][6]
        idx += 1
    end

    return starter, insertion_rules
end


def part1and2(starter, insertion_rules, steps)
    monomer_counter = {}
    for idx in 0..(starter.length() - 1)
        if monomer_counter.key?(starter[idx])
            monomer_counter[starter[idx]] += 1
        else
            monomer_counter[starter[idx]] = 1
        end
    end

    polymer_counter = {}
    for idx in 0..(starter.length() - 2)
       if polymer_counter.key?("#{starter[idx]}#{starter[idx + 1]}")
        polymer_counter["#{starter[idx]}#{starter[idx + 1]}"] += 1
       else
        polymer_counter["#{starter[idx]}#{starter[idx + 1]}"] = 1
       end
    end

    for step in 1..steps
        tmp_counter = {}
        polymer_counter.each do |key, value|
            if tmp_counter.key?("#{key[0]}#{insertion_rules[key]}")
                tmp_counter["#{key[0]}#{insertion_rules[key]}"] += value
            else
                tmp_counter["#{key[0]}#{insertion_rules[key]}"] = value
            end
            if tmp_counter.key?("#{insertion_rules[key]}#{key[1]}")
                tmp_counter["#{insertion_rules[key]}#{key[1]}"] += value
            else
                tmp_counter["#{insertion_rules[key]}#{key[1]}"] = value
            end
            if monomer_counter.key?("#{insertion_rules[key]}")
                monomer_counter["#{insertion_rules[key]}"] += value
            else
                monomer_counter["#{insertion_rules[key]}"] = value
            end
        end
        polymer_counter = tmp_counter
    end

    answer = monomer_counter.values.minmax
    return (answer[1] - answer[0])
end


starter, insertion_rules = readto_string_hash
puts "Part 1: #{part1and2(starter, insertion_rules, 10)}"
puts "Part 2: #{part1and2(starter, insertion_rules, 40)}"
__END__
